### DISCUSSION TITLE
Create collections in code

### DISCUSSION THREAD

**Discussion by tandem97** [pocketbase/pocketbase#125] [Q&A]
Hellow! I'm just starting to get familiar with this framework. It is really very usefull. I read docs on the site but did not find answers on some questions. So if someone can help me, i will be very thankful. Is there any example how i can create collections in code? When i am using dbs, my apps on start check tables and if they are not existing, app creates them  itself. I want to do the same with pocketbase. Thanks!

**Comment by ganigeorgiev** [comment]
@tandem97 Hi, this is not documented clearly yet. After the Flutter SDK I'll try to spend some time on documenting more advanced usage scenarios and will add some references.

For now, there are 2 ways to create collection in code.

### Directly using the collection model
(no validations are performed)

```go
import (
	"github.com/pocketbase/pocketbase/models"
	"github.com/pocketbase/pocketbase/models/schema"
)

...

userRequiredRule = "@request.user.id != nul"

collection := &models.Collection{
    Name:       "myNewCollection",
    System:     true,
    CreateRule: &userRequiredRule,
    ListRule:   &userRequiredRule,
    ViewRule:   &userRequiredRule,
    UpdateRule: &userRequiredRule,
    CreateRule: &userRequiredRule,
    DeleteRule: nil, // only admins will be able to access
    Schema: schema.NewSchema(
        &schema.SchemaField{
            Name:     "title",
            Type:     schema.FieldTypeText,
            Unique:   true,
            Required: true,
            System:   true, // when true, cannot be renamed or deleted
            Options:  &schema.TextOptions{},
        },
        // ... other fields
    ),
}

err := app.Dao().SaveCollection(collection)
```

### Using the `CollectionUpsert` form model
(perform validations; suitable also for loading data from requests)

```go
import (
	"github.com/pocketbase/pocketbase/forms"
	"github.com/pocketbase/pocketbase/models"
	"github.com/pocketbase/pocketbase/models/schema"
)

...

collection := &models.Collection{}
// or if you want to update existing
collection, _ := app.Dao().FindCollectionByName("example")

form := forms.NewCollectionUpsert(app, collection)
form.Name = "test"
form.Schema.AddField(&schema.SchemaField{
    Name:   "title",
    Type:     schema.FieldTypeText,
    Unique:   true,
    Required: true,
    System:   true, // when true, cannot be renamed or deleted
    Options: &schema.TextOptions{},
})

// set more fields and rules...

// save
err := form.Submit()
```

My best recommendation for now, would be to rely on the [source code documentation](https://pkg.go.dev/github.com/pocketbase/pocketbase#section-directories) and your editor LSP/autocomplete.

**Reply by tandem97** [reply]
@ganigeorgiev Thank you for your answer i will try it! Source code  documentation is great, but when framework is big and you did not work with it before, it can be difficult to find answers on some questions you have. Is there any examples of common scenarios exept examples in docs on site?

**Reply by ganigeorgiev** [reply]
Not at the moment, sorry. I completely agree that source code documentation on its own is not enough. Guides and examples with common and more advanced scenarios will be added in the near future.

**Comment by zzaba** [comment]
Is it possible to initialize some collections on start?

```
package main

import (
	"log"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/models"
	"github.com/pocketbase/pocketbase/models/schema"
)

func main() {
	app := pocketbase.New()

	if err := app.Bootstrap(); err != nil {
		log.Fatal(err)
	}

	if err := app.Dao().SaveCollection(&models.Collection{...}); err != nil {
		log.Fatal(err)
	}

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

Results in following error:

```
[0.00ms] SELECT {{_params}}.* FROM `_params` WHERE `key`='settings' LIMIT 1
[0.00ms] INSERT INTO `_collections` (`createRule`, `created`, `deleteRule`, `id`, `listRule`, `name`, `schema`, `system`, `updateRule`, `updated`, `viewRule`) VALUES (...)
2022/10/01 20:33:46 no such table: _collections
exit status 1
```

However if I first comment out the `SaveCollection` code, run it, then uncomment and run again it works.

**Reply by ganigeorgiev** [reply]
> Is it possible to initialize some collections on start?

Yes. It doesn't work in your case because the application is not bootstrapped yet. You need to wrap it in a [`OnBeforeServe` event hook](https://pocketbase.io/docs/event-hooks/#app-hooks), but keep in mind that this code will be executed every time when your application starts.

If you want to execute a certain task only once, a [DB migration](https://pocketbase.io/docs/db-migrations/) is better for this case (if you need access to the Dao service, you can initialize it inside a db migration by doing `dao := daos.New(db)`. Additionally if you are looking for Collection migration, there is also an automated command that will export your current collections configuration in a single migration - https://pocketbase.io/docs/db-migrations/#collections-migration

**Reply by zzaba** [reply]
I tried making a migration like this:

```
package migrations

import (
    "github.com/pocketbase/dbx"
    "github.com/pocketbase/pocketbase/daos"
    "github.com/pocketbase/pocketbase/models"
    "github.com/pocketbase/pocketbase/models/schema"
    m "github.com/pocketbase/pocketbase/migrations"
)

var collection = &models.Collection{
	BaseModel: models.BaseModel{
		Id: security.RandomStringWithAlphabet(models.DefaultIdLength, models.DefaultIdAlphabet),
	},
	...
}

func init() {
    m.Register(func(db dbx.Builder) error {
        return daos.New(db).SaveCollection(collection)
    }, func(db dbx.Builder) error {
        return daos.New(db).DeleteCollection(collection)
    })
}
```

But it results in following error on up migration:

```
Failed to apply migration games.go: sql: no rows in result set
```

If I comment out the `BaseModel` code, another error appears on down migration:

```
Failed to revert migration games.go: ID is not set
```

**Reply by ganigeorgiev** [reply]
This will be documented better soon in the new docs, but for now - when manually setting an ID, you have to mark the model as new (aka. collection.MarkAsNew()), but it is not ok to set a random id this way, since it will get a new value on app restart and your "down" function will not work anymore. Instead you can type a fixed id or don't set an id at all.

```go
package migrations

import (
    "github.com/pocketbase/dbx"
    "github.com/pocketbase/pocketbase/daos"
    "github.com/pocketbase/pocketbase/models"
    "github.com/pocketbase/pocketbase/models/schema"
    m "github.com/pocketbase/pocketbase/migrations"
)

const collectionId = "somerandomid123"

func init() {
    m.Register(func(db dbx.Builder) error {
        collection := &models.Collection{
            Name: "test"
            ...
        }
        collection.Id = collectionId
        collection.MarkAsNew()

        return daos.New(db).SaveCollection(collection)
    }, func(db dbx.Builder) error {
        dao := daos.New(db)

        collection, err := dao.FindCollectionByNameOrId(collectionId)
        if err != nil {
            return err
        }

        return dao.DeleteCollection(collection)
    })
}
```

But if you want to export your collections into a migration, it is better to use the `migrate collections` command.

**Reply by zzaba** [reply]
Thanks!

So the preferred way of working on schemas is the UI?

**Reply by ganigeorgiev** [reply]
I guess it depends, but for now this seems to be the easiest flow when using PocketBase as framework:

1. Make changes locally to your collections from the Admin UI
2. Run the `migrate collections` command
3. Build (`go build`)
4. Deploy to prod
5. Repeat...

_Some users reported that are not very satisfied with this, since it doesn't work when you use the prebuilt executable and I'll think how to optimize it after the upcoming v0.8.0 release and the new documentation._

**Reply by zzaba** [reply]
Yeah, I also prefer to declare my schemas in code or separate files. It'd also be cool to be able to create collection schema from protobuf / flatbuffers files.

**Reply by zzaba** [reply]
Also, since the `Id` field is the source of trouble in this case, it made me wonder, is it really necessary when we have names? I think collection and field names should be unique and so can take on whatever function `Id` field has. As in relations for example, we can specify collection name instead of id.

**Reply by ganigeorgiev** [reply]
@zzaba Collections can be renamed from the Admin UI and the `id` helps by providing a "stable" identifier since it cannot change after model create (_except with a raw sql statement_).

**Reply by H1Mohamed** [reply]
is it  possible to do the same functionality but with javascript or typescript?

**Reply by ganigeorgiev** [reply]
@H1Mohamed This is an old discussion and several things have changed since then. For creating collections with JS programattically as part of `pb_hooks` or `pb_migrations` you can check https://pocketbase.io/docs/js-collections/.
For general overview of how to write manual migrations please check https://pocketbase.io/docs/js-migrations/.

**Reply by gleb0g** [reply]
The error appears to be due to the fact that the `_collection` table is not present in the database when the `SaveCollection` method is called. After the first successful start, it is created and persists in the database file `pb_data/data.db`, which is why the error does not appear on the second run.

However, I can see the only hacky way to initialize collections on start programmatically here - put the `SaveCollection` method call in the [`OnBeforeServe`](https://pocketbase.io/docs/go-event-hooks/#onbeforeserve) hook.

```go
func main() {
    app := pocketbase.New()

    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        return app.Dao().SaveCollection(collection)
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

**Reply by ganigeorgiev** [reply]
@gleb0g Your issue doesn't seem to be related to the above discussion. My recommendation from the [previous comment](https://github.com/pocketbase/pocketbase/discussions/125#discussioncomment-6903982) remains, aka. use the migrations if you want to create collections programmatically:
- Go: https://pocketbase.io/docs/go-migrations/
- JS: https://pocketbase.io/docs/js-migrations/

If you have other questions, please consider opening a new Q&A discussion with more details about your use case and what you are trying to do.

**Reply by gleb0g** [reply]
@ganigeorgiev I apologize for the confusion. I had meant no any additional issues. My previous comment was made just as a response for initial **zzaba**'s [question](https://github.com/pocketbase/pocketbase/discussions/125#discussioncomment-3778833) in this thread.

But now I see, that I've must overlooked your first [comment](https://github.com/pocketbase/pocketbase/discussions/125#discussioncomment-3778863) which was exactly about the same thing. I'm sorry for that, I should've been more attentive

Thank you for your time and clarification 🙏
